File: December 2024 Leetcode Solution/December-01.py
# from typing import List

# class Solution:
#     def checkIfExist(self, arr: List[int]) -> bool:
#         arr.sort()  
        
#         for i, num in enumerate(arr):
#             target = num * 2
#             left = i + 1 if target >= 0 else 0
#             right = len(arr) - 1
            
#             while left <= right:
#                 mid = left + (right - left) // 2
#                 if arr[mid] == target:
#                     return True
#                 elif arr[mid] < target:
#                     left = mid + 1
#                 else:
#                     right = mid - 1
#         return False


from typing import List

class Solution:
    def checkIfExist(self, arr: List[int]) -> bool:
        seen = set()
        
        for num in arr:
            if num * 2 in seen:
                return True
            if num % 2 == 0 and num // 2 in seen:
                return True
            seen.add(num)
        
        return False

File: December 2024 Leetcode Solution/December-01.cpp
class Solution {
public:
    bool checkIfExist(vector<int>& arr) {
        sort(arr.begin(), arr.end()); 
        
        for (int i = 0; i < arr.size(); ++i) {
            int target = arr[i] * 2;
            int left = (target >= 0 ? i + 1 : 0);
            int right = arr.size() - 1;
            
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (arr[mid] == target) {
                    return true;
                } else if (arr[mid] < target) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return false;
    }
};
